Use lang_tester

ltratt opened the pull request:
Needs https://github.com/softdevteam/ykbf/pull/3 and https://github.com/softdevteam/lang_tester/pull/64 to be merged first (and a new release of lang_tester to be made).

ltratt commented:
Will need rebasing against master before merging.

ltratt commented:
bors try

ltratt commented:
All that work with fm and lang_tester seems to have paid off :) This is ready for review. I'll need to rebase away that first commit before any merge though.

vext01 commented on Cargo.toml:
```diff
@@ -5,9 +5,16 @@ authors = ["Laurence Tratt <laurie@tratt.net>"]
 
 [dev-dependencies]
-tempfile = "3.0"
+lazy_static = "1.3"
```
You could (optionally) run `cargo update`. For example lazy_static 1.4 is available (although I think it's semantically compatible with 1.3 anyway?)

ltratt replied:
`cargo update` won't do anything here, though `cargo outdated` would. Fixed in https://github.com/softdevteam/ykbf/pull/4/commits/dba29da3bde15aba3b42054e36e59ef1dd8043e1.

vext01 replied:
Your are indeed correct.

vext01 commented on lang_tests/obscure.bf:
```diff
@@ -0,0 +1,8 @@
+    stdout: H
+]
+
+[ From http://www.hevanet.com/cristofd/brainfuck/tests.b ]
```
Do we have to add any licenses if we are taking code from elsewhere?

ltratt replied:
This is a bit tricky because that site has no license. However, the code snippet is so small that I believe it comes under the exception for fair use of quotations and, as suggested by various people, we have explicitly mentioned where the quote came from.

vext01 replied:
That's fair enough. Just wanted to be sure.

vext01 commented on lang_tests/run.rs:
```diff
@@ -0,0 +1,42 @@
+            // calling `cargo run`.
+            let mut yksom_bin = PathBuf::new();
+            yksom_bin.push(env::var("CARGO_MANIFEST_DIR").unwrap());
+            yksom_bin.push("target");
```
Can we use `CARGO_BIN_EXE_ykbf` here? I've never used it, but sounds like it could make this code a lot shorter:

https://doc.rust-lang.org/cargo/reference/environment-variables.html

ltratt replied:
I did not know about that variable! Fixed in https://github.com/softdevteam/ykbf/pull/4/commits/936266e6308294e7a3b3c1f6a4a7e72e562cdc2a.

vext01 commented on lang_tests/run.rs:
```diff
@@ -0,0 +1,42 @@
+        .test_cmds(|p| {
+            // We call target/[debug|release]/yksom directly, because it's noticeably faster than
+            // calling `cargo run`.
+            let mut yksom_bin = PathBuf::new();
```
yksom? That's a copy pasta surely?

ltratt replied:
Mea culpa! Fixed in https://github.com/softdevteam/ykbf/pull/4/commits/936266e6308294e7a3b3c1f6a4a7e72e562cdc2a.

vext01 commented:
Just a few comments. Nothing significant :)

vext01 commented:
Please squasharoo.

ltratt committed:
Port tests to use lang_tester.

This isn't a big benefit at the moment, but it will make future work with ykbf a
little easier.

ltratt commented:
Squashed.

vext01 commented:
bors r+

ltratt commented:
I'm assuming that one's a bit random and just needs a retry?

vext01 commented:
Let's try once more.

bors r+
